File: src/com/tzavi/fileorganizer/services/SorterService.java
package com.tzavi.fileorganizer.services;

import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.StandardCopyOption;
import java.util.stream.Stream;

import com.tzavi.fileorganizer.config.RuleManager;
import com.tzavi.fileorganizer.ui.MenuManager;

public class SorterService {
    public static void moveFile(Path originalFile, String targetFolderName, Path targetPath) {
        Path target = targetPath.resolve(targetFolderName);
        try {
            Files.createDirectories(target);
            Path finalPath = target.resolve(originalFile.getFileName());
            Files.move(originalFile, finalPath, StandardCopyOption.REPLACE_EXISTING);
        } catch (IOException e) {
            System.out.println("Could not move file: " + originalFile.getFileName() + " (File might be in use)");
        }
    }

    public static void executeOrganization(Path folderPath) {
        if (folderPath == null || !Files.exists(folderPath)) {
            System.out.println("Invalid Path! Cannot organize.");
            return;
        }
        MenuManager.cleanscreen();
        final Path finalPath = folderPath;
        try (Stream<Path> stream = Files.list(folderPath)) {
            stream.forEach(file -> {
                if (Files.isRegularFile(file)) {
                    String fileName = file.getFileName().toString();
                    int dotIndex = fileName.lastIndexOf(".");
                    if (dotIndex > 0) {
                        String extension = fileName.substring(dotIndex + 1).toLowerCase();

                        if (RuleManager.rules.containsKey(extension)) {
                            String destinationFolder = RuleManager.rules.get(extension);
                            moveFile(file, destinationFolder, finalPath);
                        }
                    }
                }
            });
            MenuManager.filesOrganizedMessage();
        } catch (Exception e) {
            MenuManager.cleanscreen();
            System.out.println("Something went wrong when trying to organize files.");
        }
    }

    public static Path autoFindDownloadsPath() {
        Path userHome = Path.of(System.getProperty("user.home")).resolve("Downloads");
        return userHome;
    }

    public static Path manualPath() {
        MenuManager.cleanscreen();
        System.out.println("==================================================");
        System.out.println("Manually Type Path Below: ");
        System.out.println("==================================================");
        String userHomeInput = MenuManager.scanner.nextLine().trim();
        Path choosenPath = Path.of(userHomeInput).toAbsolutePath();
        if (userHomeInput.isEmpty() || !Files.exists(choosenPath)) {
            MenuManager.cleanscreen();
            System.out.println("==================================================");
            System.out.println("              Path not found");
            System.out.println("         Press ENTER to continue:");
            System.out.println("==================================================");
            MenuManager.scanner.nextLine();
            return null;
        } else {
            return choosenPath;
        }
    }

    public static Path getPath() {
        while (true) {
            try {
                MenuManager.cleanscreen();
                System.out.println("==================================================");
                System.out.println("                 Choose a path");
                System.out.println("==================================================");
                System.out.println("1 - Try Auto Find Donwloads Path\n2 - Manually Type Path");
                System.out.println("==================================================");
                String userInputString = MenuManager.scanner.nextLine();
                int userInputInt = Integer.parseInt(userInputString);
                if (userInputInt == 1) {
                    return autoFindDownloadsPath();
                } else if (userInputInt == 2) {
                    return manualPath();
                } else {
                    continue;
                }
            } catch (NumberFormatException e) {
                System.out.println("==================================================");
                System.out.println("Invalid option!\nPlease type a NUMBER");
                System.out.println("==================================================");
                MenuManager.waitingForInput();
            }
        }
    }

    public static void executeOrganizationByExtension(Path folderPath, String extension, String folderName) {
        if (folderPath == null || !Files.exists(folderPath)) {
            System.out.println("Invalid Path! Cannot organize.");
            return;
        }
        MenuManager.cleanscreen();
        try (Stream<Path> stream = Files.list(folderPath)) {
            stream.filter(file -> file.getFileName().toString().toLowerCase().endsWith("." + extension))
                    .forEach(file -> {
                        moveFile(file, folderName, folderPath);
                    });
            MenuManager.filesOrganizedMessage();
        } catch (Exception e) {
            MenuManager.cleanscreen();
            System.out.println("Something went wrong when trying to organize files.");
        }
    }
}

File: Services/SorterService.cs
using FileOrganizer.FileOrganizer.Config;
using FileOrganizer.FileOrganizer.UI;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace FileOrganizer.FileOrganizer.Services
{
    internal class SorterService
    {
        public static string GetPath()
        {
            while (true)
            {
                MenuManager.CleanScreen();
                Console.WriteLine("=====================================================");
                Console.WriteLine("                 Choose a path");
                Console.WriteLine("=====================================================");
                Console.WriteLine("1 - Try Auto Find Donwloads Path\n2 - Manually Type Path");
                Console.WriteLine("=====================================================");
                string userInputString = Console.ReadLine();
                if (int.TryParse(userInputString, out int userInputInt))
                {
                    switch (userInputInt)
                    {
                        case 1:
                            return AutoFindDownloadsPath();
                        case 2:
                            return GetManualPath();
                    }
                }
                Console.WriteLine("=====================================================");
                Console.WriteLine("Invalid option!\nPlease Type a valid option");
                Console.WriteLine("=====================================================");
                MenuManager.WaitingForInput();
            }
        }

        public static string AutoFindDownloadsPath()
        {
            string downloadsPath = Path.Combine(Environment.GetFolderPath(Environment.SpecialFolder.UserProfile), "Downloads");
            return downloadsPath;
        }

        public static string GetManualPath()
        {
            MenuManager.CleanScreen();
            Console.WriteLine("=====================================================");
            Console.WriteLine("           Manually Type Path Below: ");
            Console.WriteLine("=====================================================");
            String path = Console.ReadLine();
            if (!Path.Exists(path) || string.IsNullOrWhiteSpace(path))
            {
                MenuManager.CleanScreen();
                Console.WriteLine("=====================================================");
                Console.WriteLine("                  Path not found");
                Console.WriteLine("            Press ENTER to try continue:");
                Console.WriteLine("=====================================================");
                Console.ReadLine();
                return null;
            }
            else
            {
                return path;
            }
        }

        public static void MoveFile(string originalFile, string folderName, string path)
        {
            string targetFolder = Path.Combine(path, folderName);
            Directory.CreateDirectory(targetFolder);
            string fileNameOnly = Path.GetFileName(originalFile);
            string finalPath = Path.Combine(targetFolder, fileNameOnly);
            File.Move(originalFile, finalPath, true);
        }

        public static void ExecuteOrganization(string path)
        {
            if (path == null || !Path.Exists(path))
            {
                Console.WriteLine("Invalid Path! Cannot organize.");
                return;
            }
            string[] files = Directory.GetFiles(path);
            MenuManager.CleanScreen();
            foreach (string file in files)
            {
                string extension = Path.GetExtension(file).TrimStart('.');
                if (RuleManager.rules.ContainsKey(extension))
                {
                    string destinationFolder = RuleManager.rules[extension];
                    MoveFile(file, destinationFolder, path);
                }
            }
            MenuManager.FilesOrganizedMessage();
        }

        public static void ExecuteOrganizationByExtension(string path, string extension, string folderName)
        {
            if (path == null || !Path.Exists(path))
            {
                Console.WriteLine("Invalid Path! Cannot organize.");
                return;
            }
            MenuManager.CleanScreen();
            var targetFiles = Directory.GetFiles(path).Where(f => f.ToLower().EndsWith("." + extension));
            foreach (var file in targetFiles)
            {
                MoveFile(file, folderName, path);
            }
        }
    }
}
